### ISSUE TITLE
ansible 2.2 doesn't support direct dictionary on task vars

### ISSUE THREAD

**Issue by joaocc** [ansible/ansible#19084]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
lib/ansible/playbook/base.py

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file =
  configured module search path = Default w/o overrides
```

##### OS / ENVIRONMENT
N/A

##### SUMMARY
I am trying to set the vars for either a block: or an include_role: from a dictionary.

I get one of these messages (depending on the case):
    ERROR! Vars in a Block must be specified as a dictionary, or a list of dictionaries
    ERROR! Vars in a IncludeRole must be specified as a dictionary, or a list of dictionaries

##### STEPS TO REPRODUCE
```
- set_fact:
    _T_m3_inv_vars:
        one_setting: "{{ some_other_var1 |mandatory }}"
        two_setting: "{{ some_other_var2 |mandatory }}"

- block:
    - debug: msg="YYYYYYYY"
  vars: "{{ _T_m3_inv_vars }}"

- include_role:
      name: "some_nice_role"
      static:  yes
      private: yes
  vars: "{{ _T_m3_inv_vars }}"

```

##### EXPECTED RESULTS
I would expect a similar result as having done the following
```
- block:
    - debug: msg="YYYYYYYY"
  vars:
        one_setting: "{{ some_other_var1 |mandatory }}"
        two_setting: "{{ some_other_var2 |mandatory }}"

- include_role:
      name: "some_nice_role"
      static:  yes
      private: yes
  vars:
        one_setting: "{{ some_other_var1 |mandatory }}"
        two_setting: "{{ some_other_var2 |mandatory }}"
```
This is because I am indeed passing a dictionary to the vars: setting.

##### ACTUAL RESULTS
One of these messages (obtained by running the previous snippets separatellyI
    ERROR! Vars in a Block must be specified as a dictionary, or a list of dictionaries
    ERROR! Vars in a IncludeRole must be specified as a dictionary, or a list of dictionaries

**Comment by mprasil** [comment]
To dumb down the steps to reproduce this playbook ends up with the _"ERROR! Vars in a TaskInclude must be specified as a dictionary, or a list of dictionaries"_ error:

```yml
- hosts: localhost
  tasks:
  - include: "inc.yml"
    vars: "{{ {'a': 1} }}"
```

While this one works fine:

```yml
- hosts: localhost
  tasks:
  - include: "inc.yml"
    vars: {'a': 1}
```

I was able to reproduce this with Ansible version 2.2.1.0.

**Comment by mprasil** [comment]
There's a possible workaround if you can change the role/task file you're including. You can set some hand-over variable like this:

```yml
# playbook.yml
- hosts: localhost
  tasks:
  - include: "inc.yml"
    vars:
        handover: "{{ {'a': 1 } }}"
```

and then as a first step in your included file:

```yml
# inc.yml
- set_fact:
  args: "{{ handover }}"
  when: handover
```

This throws a deprecation warning, but works for now. Hopefully this will be fixed before the functionality is deprecated.

Hope that helps.

**Comment by jurgenweber** [comment]
yeah, I am getting this in 2.3 as well.

**Comment by bcoca** [comment]
The 'vars' type is being validated PRE templating so any template looks like a string at that point.

Just to point out,  ```  vars: "{{ _T_m3_inv_vars }}"```  just copies already accessible data, the included file can already get to.

**Comment by krzysztof-magosa** [comment]
@bcoca I would like to create wrapper to existing playbook to run it many times with parameters from list. That way I would keep role simple without tons of loops etc, and just few lines of wrapper would do a job. Do you plan changing code to validate POST templating or it won't happen?

**Comment by jstockwin** [comment]
We've run into this problem too, however the solution in [mprasil's comment](https://github.com/ansible/ansible/issues/19084#issuecomment-289323733) did not work for us, I think because using `args` has now been deprecated and no longer works (we're using 2.5.0).

Therefore, instead of adding the following as @mprasil suggests:
```
# inc.yml
- set_fact:
  args: "{{ handover }}"
  when: handover
```
we've done:
```
- set_fact: {"{{ item.key }}": "{{ item.value }}"}
  with_dict: "{{ handover }}"
  when: handover is defined
```
which achieves the same result.

**Comment by sivel** [comment]
More basic reproducer:
```yaml
- hosts: localhost
  gather_facts: false
  tasks:
    - debug: var=a
      vars: "{{ {'a': 1} }}"
```

**Comment by sivel** [comment]
!component =lib/ansible/playbook/base.py

**Comment by krzysztof-magosa** [comment]
Any progress guys?

**Comment by juanvalino** [comment]
+1 this feature is a must if we want real meta-programming in ansible.

**Comment by seblu** [comment]
This feature seems to be mandatory to mimic the support of [dictionary roles](https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html#using-roles) in traditional ansible playbook with a dynamic list of roles defined in inventory/host_vars/groups_vars.

##### Example of host_vars/hostname.yml
```
roles:
  - base
  - role: gitlab
    foo: bar
```

##### Example of site.yml
```
  tasks:
    - name: run {{ ansible_host }} roles
      include_role:
        name: "{{ roleloop is mapping|ternary(roleloop.role, roleloop) }}"
        apply:
          tags:
            - "{{ roleloop.role|default(roleloop) }}"
            - "{{ roleloop.tags|default(omit) }}"
      vars: "{{ roleloop if roleloop is mapping else {} }}"
      loop: "{{ roles|default([]) }}"
      loop_control:
        loop_var: roleloop
      tags:
        - always
```

**Comment by ragingpastry** [comment]
> This feature seems to be mandatory to mimic the support of [dictionary roles](https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html#using-roles) in traditional ansible playbook with a dynamic list of roles defined in inventory/host_vars/groups_vars.
> 
> ##### Example of host_vars/hostname.yml
> ```
> roles:
>   - base
>   - role: gitlab
>     foo: bar
> ```
> ##### Example of site.yml
> ```
>   tasks:
>     - name: run {{ ansible_host }} roles
>       include_role:
>         name: "{{ roleloop is mapping|ternary(roleloop.role, roleloop) }}"
>         apply:
>           tags:
>             - "{{ roleloop.role|default(roleloop) }}"
>             - "{{ roleloop.tags|default(omit) }}"
>       vars: "{{ roleloop if roleloop is mapping else {} }}"
>       loop: "{{ roles|default([]) }}"
>       loop_control:
>         loop_var: roleloop
>       tags:
>         - always
> ```

I am attempting to do this exact same thing and am running into the same issue. In my opinion, it is much easier and cleaner to define the roles and role variables in the same place. 

```
common_roles:
  - name: security
    vars:
      - security_selinux_enable: True
      - security_firewalld_enable True
```

This format makes much more sense because it allows me to keep the variables and the roles in the same place.

**Comment by chrrrles** [comment]
> +1 this feature is a must if we want real meta-programming in ansible.

Currently trying to drive ansible plays using a "manifest" that is parsed by an inventory script - the script defines roles applied for each host, and also the variables needed by the roles.  Been trying to wrap my head around a solution to this problem without hacking Ansible.

I understand that Ansible evaluates `vars:` before the play so there is no way to dynamically feed a dictionary to an `include_role` with that.  It may be worth extending include_role to support dynamic variables.

**Comment by jamescassell** [comment]
My reproducer is:
```
- include_role:
  - name: my-role
  vars: "{{ my_role_vars }}"
```
which results in `ERROR! Vars in a IncludeRole must be specified as a dictionary, or a list of dictionaries` at which point, I tried:
```
- include_role:
  - name: my-role
  vars:
  - "{{ my_role_vars }}"
```
which resulted in the same error.

I also tried:
```
- include_role:
  - name: my-role
    apply:
      vars:
      - "{{ my_role_vars }}"
```
Same error.

**Comment by sivel** [comment]
We have decided in the most recent core irc meeting to not pursue this.  On top of not being completely necessary, this would cause further confusion about variable precedence. 

When trying to use this for roles, it is recommended that the role be re-designed to work with looping in mind, exposing the ability for the role to use a top level variable such as `item`.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Comment by ragingpastry** [comment]
I'm not sure I agree with this causing confusion with variable precedence. Can you expand on what you mean by that? 

Also I would argue that within the context of this issue it is necessary and logical to define variables like this.

**Comment by jamescassell** [comment]
The only workaround I've been able to come up with is to dynamically write a tasks file then include_tasks that file. It's simply too invasive to rewrite every role to natively support looping.

**Comment by ragingpastry** [comment]
> The only workaround I've been able to come up with is to dynamically write a tasks file then include_tasks that file. It's simply too invasive to rewrite every role to natively support looping.

I've been using an include_task which calls a task with this:

`- name: Set role variable facts
  set_fact:
    "{{ role_item.key }}": "{{ role_item.value }}"
  with_dict: "{{ item.vars }}"
  loop_control:
    loop_var: role_item`

It works but its messy...

**Comment by jamescassell** [comment]
set_fact works until you do a later group_by that you'd expect to pull in more specific group_vars...

**Comment by ragingpastry** [comment]
Oh indeed. It fit my use-case for the moment but doesn't solve the problem which is why I was a little concerned that this was closed as "not necessary"

**Comment by smemsh** [comment]
Great, so now we've permanently declared `include_role` as a "special" tasks which works differently than other tasks, that take variables in vars just fine (eg `debug`).  Even the people making the tests know it's broken, as in:
https://github.com/ansible/ansible/blob/cf6c7658714402ff320e111d268fb688d23ae0b0/test/integration/targets/include_import/role/test_include_role.yml#L99-L105

It says FIXME because it's *broken*.  Making every role need special knowledge about external loop vars and have some strange convention built into every role just to pass arguments is a clear abstraction violation, and breaks separation of concerns.  Not to mention working differently than common sense, and than other modules do.

It took several years before loop variables could be used as `name` arguments to `include_role`.  Now that it's half complete, we'll apparently never get to finish the job and pass parameters to them.  So much for encapsulation... roles will never be first class citizens in Ansible.

**Comment by jamescassell** [comment]
So, I decided it was too much work to rewrite every role to support passing all parameters in a single variable.  I ended up writing this hack of a role that allows the same thing, but in a much more ugly way: https://github.com/jamescassell/ansible-role-loop

It doesn't (yet) support being called from within itself, but otherwise should work for most generic roles.  The only hang ups I've had is where a role doesn't expect to be called more than once and its handlers only clean up resources from the most recent call, but that is something the role should handle if it wants to be called multiple times.

As long as you don't look inside the above role, it's almost as if ansible itself supports the feature.
